### PR TITLE
KV: do not use "Inputs" part of InputsAndEffects

### PR DESCRIPTION
**Pull request by remyhaemmerle-da**
CHANGELOG_BEGIN
CHANGELOG_END

### Pull Request Checklist

- [ ] Read and understand the [contribution guidelines](https://github.com/digital-asset/daml/blob/main/CONTRIBUTING.md)
- [ ] Include appropriate tests
- [ ] Set a descriptive title and thorough description
- [ ] Add a reference to the [issue this PR will solve](https://github.com/digital-asset/daml/issues), if appropriate
- [ ] Include changelog additions in one or more commit message bodies between the `CHANGELOG_BEGIN` and `CHANGELOG_END` tags
- [ ] Normal production system change, include purpose of change in description

NOTE: CI is not automatically run on non-members pull-requests for security
reasons. The reviewer will have to comment with `/AzurePipelines run` to
trigger the build.

### CODE REVIEW

**Review by miklos-da — COMMENTED** [review]
@remyhaemmerle-da If you merge #9361 you won't need this PR. Am I missing something?

**Comment by remyhaemmerle-da** [comment]
> @remyhaemmerle-da If you merge #9361 you won't need this PR. Am I missing something?

I think it is better to go this way. Review #9361 and make it only about Effect and make this PR about Inputs only.

In the concrete case we do not need a special computation of Input because almost everything (expect contract key) is already in GenTransaction. 

I tried to address some of the comment you make in #9361, here.

File: daml-lf/transaction/src/main/scala/com/digitalasset/daml/lf/transaction/Transaction.scala [hunk 1]
```diff
@@ -384,6 +384,24 @@ sealed abstract class HasTxNodes[Nid, +Cid] {
       case (acc, _) => acc
     } -- localContracts.keySet
 
+  /** Return all the contract keys created, exercised, fetched or lookup by this transation.
+    */
+  final def contractKeys[Cid2 >: Cid]: Set[(Ref.Identifier, Value[Cid2])] = {
```
Comment (cocreature) [inline daml-lf/transaction/src/main/scala/com/digitalasset/daml/lf/transaction/Transaction.scala:389]: Should we have a test for this somewhere?

File: daml-lf/transaction/src/main/scala/com/digitalasset/daml/lf/transaction/Transaction.scala [hunk 2]
```diff
@@ -384,6 +384,24 @@ sealed abstract class HasTxNodes[Nid, +Cid] {
       case (acc, _) => acc
     } -- localContracts.keySet
 
+  /** Return all the contract keys created, exercised, fetched or lookup by this transation.
```
Comment (cocreature) [inline daml-lf/transaction/src/main/scala/com/digitalasset/daml/lf/transaction/Transaction.scala:387]: Can we be a bit more precise here? What exactly does `all` mean? Does it include transient contracts, created and archived within this transaction? That probably then also defines what we do with rollback nodes.

File: ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/participant/state/kvutils/KeyValueSubmission.scala [hunk 3]
```diff
@@ -44,14 +46,28 @@ class KeyValueSubmission(metrics: Metrics) {
...
+      val packageIdStates = meta.optUsedPackages
+        .getOrElse(
+          throw new InternalError("Transaction was not annotated with used packages")
+        )
+        .map(Conversions.packageStateKey)
+      val partyStates =
+        (tx.informees ++ submitterInfo.actAs).toList.map(Conversions.partyStateKey)
+      val contractIdStates =
+        tx.inputContracts[ContractId].map(Conversions.contractIdToStateKey)
+      val contractKeyStates =
```
Comment (cocreature) [inline ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/participant/state/kvutils/KeyValueSubmission.scala:59]: Why do we treat contract keys created with the transaction as inputs but don’t do the same for contract ids? That seems a bit inconsistent. I assume it has something to do with duplicate key detection or something along those lines?

File: daml-lf/transaction/src/main/scala/com/digitalasset/daml/lf/transaction/Transaction.scala (hunk 1 above)
Comment (cocreature) [reply daml-lf/transaction/src/main/scala/com/digitalasset/daml/lf/transaction/Transaction.scala:389]: fine with merging without tests given that there don’t really seem to be tests for the inputs and effects logic either but we should fix this so maybe add a todo.

**Review by cocreature — APPROVED** [review]

File: ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/participant/state/kvutils/KeyValueSubmission.scala (hunk 3 above)
Comment (remyhaemmerle-da) [reply ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/participant/state/kvutils/KeyValueSubmission.scala:59]: As far I understand, for validation, KV have to collect all the info needed to replay the transaction. For created contract it should know the contract key does not exists.

Comment (cocreature) [reply ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/participant/state/kvutils/KeyValueSubmission.scala:59]: alright so the input is the “non-existence of the key” or something like that, makes sense :+1:

File: daml-lf/transaction/src/main/scala/com/digitalasset/daml/lf/transaction/Transaction.scala (hunk 1 above)
Comment (remyhaemmerle-da) [reply daml-lf/transaction/src/main/scala/com/digitalasset/daml/lf/transaction/Transaction.scala:391]: done

File: daml-lf/transaction/src/main/scala/com/digitalasset/daml/lf/transaction/Transaction.scala (hunk 2 above)
Comment (remyhaemmerle-da) [reply daml-lf/transaction/src/main/scala/com/digitalasset/daml/lf/transaction/Transaction.scala]: done.

File: daml-lf/transaction/src/test/scala/com/digitalasset/daml/lf/transaction/TransactionSpec.scala [hunk 4]
```diff
@@ -268,6 +272,106 @@ class TransactionSpec extends AnyFreeSpec with Matchers with ScalaCheckDrivenPro
 
     }
   }
+
+  "contractKeys" - {
+    "return all the contract key" in {
```
Comment (remyhaemmerle-da) [inline daml-lf/transaction/src/test/scala/com/digitalasset/daml/lf/transaction/TransactionSpec.scala:277]:
```suggestion
    "return all the contract keys" in {
```

File: daml-lf/transaction-test-lib/src/main/scala/lf/transaction/test/TransactionBuilder.scala [hunk 5]
```diff
@@ -49,6 +49,8 @@ final class TransactionBuilder(
     nodes(parentId) match {
       case _: TxExercise | _: TxRollback =>
         children += parentId -> (children(parentId) :+ nodeId)
+      case _: TxRollBack =>
+        children += parentId -> (children(parentId) :+ nodeId)
```
Comment (SamirTalwar-DA) [inline daml-lf/transaction-test-lib/src/main/scala/lf/transaction/test/TransactionBuilder.scala:53]: Is this clause ever hit, given the above one?

File: daml-lf/transaction/src/main/scala/com/digitalasset/daml/lf/transaction/Transaction.scala [hunk 6]
```diff
@@ -384,6 +384,27 @@ sealed abstract class HasTxNodes[Nid, +Cid] {
...
     } -- localContracts.keySet
 
+  /** Return all the contract keys references by this transaction.
+    * This includes the keys created, exercised, fetched, or lookup, even those
+    * that refer to transient contracts or that appear under a roolback node.
+    */
+  final def contractKeys(implicit
+      ev: HasTxNodes[Nid, Cid] <:< HasTxNodes[_, Value.ContractId]
+  ): Set[GlobalKey] = {
+    ev(this).fold(Set.empty[GlobalKey]) {
```
Comment (SamirTalwar-DA) [inline daml-lf/transaction/src/main/scala/com/digitalasset/daml/lf/transaction/Transaction.scala:394]: I'd find this a little prettier with some kind of `foldMap`. Is it worth implementing that?

File: daml-lf/transaction/src/test/scala/com/digitalasset/daml/lf/transaction/TransactionSpec.scala [hunk 7]
```diff
@@ -268,6 +272,108 @@ class TransactionSpec extends AnyFreeSpec with Matchers with ScalaCheckDrivenPro
...
```
Comment (SamirTalwar-DA) [inline daml-lf/transaction/src/test/scala/com/digitalasset/daml/lf/transaction/TransactionSpec.scala:277]:
```suggestion
    "return all the contract keys" in {
```

File: daml-lf/transaction/src/main/scala/com/digitalasset/daml/lf/transaction/Transaction.scala (hunk 6 above)
Comment (cocreature) [reply daml-lf/transaction/src/main/scala/com/digitalasset/daml/lf/transaction/Transaction.scala:394]: Could be sensible but I’d prefer keeping it for a separate PR.

File: daml-lf/transaction-test-lib/src/main/scala/lf/transaction/test/TransactionBuilder.scala (hunk 5 above)
Comment (cocreature) [reply daml-lf/transaction-test-lib/src/main/scala/lf/transaction/test/TransactionBuilder.scala]: good point, I think this was a botched merge, fixed in 0667cc76a35204a2165b1395ab54d5c457376900

**Review by SamirTalwar-DA — APPROVED** [review]
I'm happy but if I were you, I'd wait for @miklos-da or @fabiotudone-da to verify. I am not _too_ familiar with this part of the code.

File: ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/participant/state/kvutils/KeyValueSubmission.scala [hunk 8]
```diff
@@ -44,14 +46,24 @@ class KeyValueSubmission(metrics: Metrics) {
...
     metrics.daml.kvutils.submission.conversion.transactionToSubmission.time { () =>
-      val inputDamlStateFromTx = InputsAndEffects.computeInputs(tx, meta)
       val encodedSubInfo = buildSubmitterInfo(submitterInfo)
+      val packageIdStates = meta.optUsedPackages
+        .getOrElse(
+          throw new InternalError("Transaction was not annotated with used packages")
+        )
+        .map(Conversions.packageStateKey)
+      val partyStates =
+        (tx.informees ++ submitterInfo.actAs).toList.map(Conversions.partyStateKey)
```
Comment (fabiotudone-da) [inline ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/participant/state/kvutils/KeyValueSubmission.scala:56]: How about encapsulating `(tx.informees ++ submitterInfo.actAs)` too somehow/somewhere?

File: daml-lf/transaction/src/test/scala/com/digitalasset/daml/lf/transaction/TransactionSpec.scala [hunk 9]
```diff
@@ -268,6 +272,108 @@ class TransactionSpec extends AnyFreeSpec with Matchers with ScalaCheckDrivenPro
 
     }
   }
+
+  "contractKeys" - {
+    "return all the contract keys" in {
+      // TODO: https://github.com/digital-asset/daml/issues/8020
+      // change VDev to  TransactionVersion.StableVersions.max once exception are released
```
[the remark this answers is not in the document]
Comment (cocreature) [reply daml-lf/transaction/src/test/scala/com/digitalasset/daml/lf/transaction/TransactionSpec.scala:279]: We generally keep these todos around so it’s easy to grep for them before closing the issue.

File: ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/participant/state/kvutils/KeyValueSubmission.scala (hunk 8 above)
Comment (cocreature) [reply ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/participant/state/kvutils/KeyValueSubmission.scala:56]: 5f95fd0b3abaf0e7715307993edd36d476ff6e8f if you have a better name for it, I’m happy to change it. I considered `partyStates` but that doesn’t seem much better.

**Review by fabiotudone-da — APPROVED** [review]

File: daml-lf/transaction/src/main/scala/com/digitalasset/daml/lf/transaction/Transaction.scala [hunk 10]
```diff
@@ -384,6 +384,27 @@ sealed abstract class HasTxNodes[Nid, +Cid] {
       case (acc, _) => acc
     } -- localContracts.keySet
 
+  /** Return all the contract keys referenced by this transaction.
+    * This includes the keys created, exercised, fetched, or looked up, even those
+    * that refer transient contracts or that appear under a rollback node.
+    */
+  final def contractKeys(implicit
+      ev: HasTxNodes[Nid, Cid] <:< HasTxNodes[_, Value.ContractId]
```
Comment (miklos-da) [inline daml-lf/transaction/src/main/scala/com/digitalasset/daml/lf/transaction/Transaction.scala:392]: What is an `ev`? Can you use meaningful parameter names, please?

Comment (SamirTalwar-DA) [reply daml-lf/transaction/src/main/scala/com/digitalasset/daml/lf/transaction/Transaction.scala:392]: `ev` is pretty conventional in Scala-ese for "evidence".

I prefer the latter, but I think it's quite common (kind of like "i" for an iterator or "e" for an exception).

Comment (cocreature) [reply daml-lf/transaction/src/main/scala/com/digitalasset/daml/lf/transaction/Transaction.scala:392]: https://github.com/digital-asset/daml/pull/9442

Comment (miklos-da) [reply daml-lf/transaction/src/main/scala/com/digitalasset/daml/lf/transaction/Transaction.scala:392]: Thanks for the explanation, @SamirTalwar-DA.

Comment (remyhaemmerle-da) [reply daml-lf/transaction/src/main/scala/com/digitalasset/daml/lf/transaction/Transaction.scala:392]: The method [Iterable#toMap](https://github.com/scala/scala/blob/8684ae833dcfeac6107343fcca5501301e509eef/src/library/scala/collection/GenTraversableOnce.scala#L648) is the standard example of method using such implicit (and syntax)

**Review by remyhaemmerle-da — COMMENTED** [review]
LGTM

**Pull request merged**